charts/opentelemetry-operator/README.md

```diff
@@ -0,0 +1,319 @@
+Consult also the [Helm Documentation on CRDs](https://helm.sh/docs/chart_best_practices/custom_resource_definitions).
+
+_See [helm upgrade](https://helm.sh/docs/helm/helm_upgrade/) for command documentation._
+
+## Configuration
+
+The following command will show all the configurable options with detailed comments.
+
+```console
+$ helm show values open-telemetry/opentelemetry-operator
+```
+
+## Install OpenTelemetry Collector
+
+_See [OpenTelemetry website](https://opentelemetry.io/docs/collector/) for more details about the Collector_
+
+Once the opentelemetry-operator deployment is ready, you can deploy OpenTelemetry Collector in our Kubernetes
+cluster.
+
+The Collector can be deployed as one of four modes: Deployment, DaemonSet, StatefulSet and Sidecar. The default
+mode is Deployment. We will introduce the benefits and use cases of each mode as well as giving an example for each.
+
+### Deployment Mode
+
+If you want to get more control of the OpenTelemetry collector and create a standalone application, Deployment would
+be your choice. With Deployment, you can relatively easily scale up the collector to monitor more targets, roll back
+to an early version if anything unexpected happens, pause the collector, etc. In general, you can manage your collector
+instance just as an application.
+
+The following example configuration deploys the Collector as Deployment resource. The receiver is jaeger-receiver and
+the exporter is logging exporter.
+
+```console
+$ kubectl apply -f - <<EOF
+apiVersion: opentelemetry.io/v1alpha1
+kind: OpenTelemetryCollector
+metadata:
+  name: my-collector
+spec:
+  mode: deployment # This configuration is omittable.
+  config: |
+    receivers:
+      jaeger:
+        protocols:
+          grpc:
+    processors:
+
+    exporters:
+      logging:
+
+    service:
+      pipelines:
+        traces:
+          receivers: [jaeger]
+          processors: []
+          exporters: [logging]
+EOF
+```
+
+### DaemonSet Mode
+
+DaemonSet should satisfy your most basic needs and is the most common method to deploy OpenTelemetry collector.
+In this case, every Kubernetes node will have its own collector copy which would monitor the pods in it.
+
+The following example configuration deploys the Collector as DaemonSet resource. The receiver is jaeger-receiver and
+the exporter is logging exporter.
+
+```console
+$ kubectl apply -f - <<EOF
+apiVersion: opentelemetry.io/v1alpha1
+kind: OpenTelemetryCollector
+metadata:
+  name: my-collector
+spec:
+  mode: daemonset
+  config: |
+    receivers:
+      jaeger:
+        protocols:
+          grpc:
+    processors:
+
+    exporters:
+      logging:
+
+    service:
+      pipelines:
+        traces:
+          receivers: [jaeger]
+          processors: []
+          exporters: [logging]
+EOF
+```
+
+### StatefulSet Mode
+If you want your collector to have stable persistent identities or storage, you should choose StatefulSet.
```

Comment:
```suggestion
If you want your Collector to have stable persistent identities or storage, you should choose StatefulSet.
```

Persistent storage isn't a promise of StatefulSets. When a pod fails, Kubernetes will _attempt_ to schedule the new pod with the same name to the node that held the pod until now. Kubernetes will also _attempt_ to mount the same volume it was used before.

This means that if the node is down, if the pod with the same name will be created in a different node. When using local storage, this means that data will be lost.

Reply:
> This means that if the node is down, if the pod with the same name will be created in a different node. When using local storage, this means that data will be lost.

I'm sorry, but can you elaborate on this? I'm not sure if I get your points. Thanks.

Reply:
Sure, let me try, hoping that I have had enough coffee this time :-) When having a stateful set with one replica, you'll have a pod named "otelcol-0". This will be scheduled to run in the node, say, "node-0".

When using a local storage like hostPath, when "otelcol-0" crashes, Kubernetes will attempt to schedule the new "otelcol-0" to the same node (node-0). If Kubernetes can't schedule the pod to this node (perhaps the node itself crashed), then a new node will be assigned to the pod. Now, "otelcol-0" is running on "node-1". Your hostPath now points to a location that has no data at all, so, your state has not been persisted.

Hence my comment that persistent storage isn't a promise of StatefulSets. They will attempt to run on the same node and attach the same volumes, but using a statefulset doesn't guarantee that you'll see the same data as before when a pod is restarted.

Reply:
Thanks for this explanation. The scenario that the node itself may crash should be considered.

Since in the Collector case, persistent storage isn't what's the StatefulSet focuses, I removed the PV related expression and underscored three advantages instead. Please LMK what's your thoughts on that.